File: DAY_08/Ravi/Unique Binary Search Trees.cpp
class Solution {
public:
    int numTrees(int n) {
        long long dp[n+1] ;
    memset(dp, 0, sizeof(dp));
    dp[0] = 1;
    dp[1] = 1;      
    
    for(long long i = 2; i <= n; i++){            
        long long l = 0, r = (i - 1);            
        while( l <= (i - 1) ) {
            dp[i] += dp[l] * dp[r];
            l++;
            r--;
        }            
    }
    return dp[n];
}
};

File: DAY_08/Ravi/Unique Binary Search Trees.java
// ## JAVA MEMORIZATION + RECURSION SOLUTION.


class Solution {
    public int trav(int l,int h,int dp[][])
    {
        if(l>h)
            return 0;
        if(l==h)
            return 1;
        if(dp[l][h]!=0)
            return dp[l][h];
        int ans=0;
        for(int i=l;i<=h;i++)
        {
            int left=trav(l,i-1,dp);
            int right=trav(i+1,h,dp);
            if(left==0 || right==0)
                ans=ans+left+right;
            else
                ans=ans+(left*right);  
        }
        dp[l][h]=ans;
        return ans;
    }
    public int numTrees(int n) {
        int dp[][]=new int[n+1][n+1];
        return trav(1,n,dp);
    }
}
